pkg/flow/internal/controller/component.go

```diff
@@ -272,6 +277,7 @@ func (cn *ComponentNode) Run(ctx context.Context) error {
 	}
 
 	cn.setRunHealth(component.HealthTypeHealthy, "started component")
+	defer cn.managedOpts.Registerer.UnregisterComponent()
```

Comment:
This has the same leaking problem as before (Run won't always get called even after metrics get registered), but now it's at the controller level.

I still think we need to implement a collector which iterates over existing ComponentNodes and collects from their registries similar to the [collector Marc wrote](https://github.com/grafana/agent/blob/1bc903abb68d4889a5742e69ae937f5e6a0a56b2/pkg/flow/internal/controller/metrics.go#L38-L70).